File: AdventOfCode2022/Python/Day11/Day11.py
def get_inputs(file):
    """
    function to read in data from a txt file
    """

    with open(file) as f:
        data = f.read().split('\n')

    return data

def MonkeyParser(data):
    """
    parse the monkey data into a useable 
    list of dictionaries
    """

    monkeys = []
    current_monkey = {}
    
    for i in range(0,len(data)):

        line = data[i]

        if line.strip() == '':
            continue

        if (i == (len(data)-1)):
            monkeys.append(current_monkey)
        
        if line.split(' ')[0] == "Monkey":
            monkeys.append(current_monkey)
            current_monkey = {}
            current_monkey['number'] = int(line.split(' ')[1][:-1])
        
        else:
            if line.strip().split(' ')[0] == "Starting":
                starting_items = line.split(':')[1].split(',')
                ints = []
                for item in starting_items:
                    ints.append(int(item))
                current_monkey['starting_items'] = ints
            
            if line.strip().split(':')[0] == "Operation":
                current_monkey['operation'] = line.split(':')[1].strip().split('=')[1]
            
            if line.strip().split(' ')[0] == "Test:":
                current_monkey['test_divisible_by'] = int(line.split(':')[1].strip().split(' ')[-1])

            if line.strip().split(':')[0] == "If true":
                current_monkey['if_true_throw_to_monkey'] = int(line.split(':')[1].strip().split(' ')[-1])
            
            if line.strip().split(':')[0] == "If false":
                current_monkey['if_false_throw_to_monkey'] = int(line.split(':')[1].strip().split(' ')[-1])
    
    return monkeys[1:]

def Part1(filename):
    """
    Solutions to Part 1
    """
    data = get_inputs(filename)

    monkeys = MonkeyParser(data)

    inspections = [0]*len(monkeys)
    for round in range(1,20 + 1):

        for m in range(0,len(monkeys)):

            starting_items           = monkeys[m]['starting_items']   
            operation                = monkeys[m]['operation']           
            test_divisible_by        = monkeys[m]['test_divisible_by']
            if_true_throw_to_monkey  = monkeys[m]['if_true_throw_to_monkey']
            if_false_throw_to_monkey = monkeys[m]['if_false_throw_to_monkey']

            for worry_level in starting_items:
                inspections[m] += 1

                old = worry_level
                worry_level = eval(operation)
                worry_level = int(worry_level / 3)

                if worry_level % test_divisible_by == 0:
                    monkeys[if_true_throw_to_monkey]['starting_items'].append(worry_level)
                else:
                    monkeys[if_false_throw_to_monkey]['starting_items'].append(worry_level)
            
            monkeys[m]['starting_items'] = []
        
    inspections.sort()
    inspections.reverse()
    return inspections[0] * inspections[1]

def Part2(filename):
    """
    Solutions to Part 2
    """
    data = get_inputs(filename)
    
    monkeys = MonkeyParser(data)
    LCM = 1
    for monkey in monkeys:
        LCM *= monkey['test_divisible_by']

    inspections = [0]*len(monkeys)
    for round in range(1,10000 + 1):
        
        for m in range(0,len(monkeys)):

            for old in monkeys[m]['starting_items']:
                inspections[m] += 1

                worry_level = eval(monkeys[m]['operation']) 

                # monkey anxiety medicine (thanks fergus)
                worry_level = worry_level % LCM

                if worry_level % monkeys[m]['test_divisible_by'] == 0:
                    monkeys[monkeys[m]['if_true_throw_to_monkey']]['starting_items'].append(worry_level)
                else:
                    monkeys[monkeys[m]['if_false_throw_to_monkey']]['starting_items'].append(worry_level)
            
            monkeys[m]['starting_items'] = []

    inspections.sort()
    inspections.reverse()
    return inspections[0] * inspections[1]

def main():
    """
    main function
    """ 
    testfile  = "tests.txt"
    inputfile = "inputs.txt"

    print("Advent-Of-Code 2022 - Day11")

    print(f'"{testfile}": Answer to Part 1 = {Part1(testfile)}')
    print(f'"{testfile}": Answer to Part 2 = {Part2(testfile)}\n')

    print(f'"{inputfile}": Answer to Part 1 = {Part1(inputfile)}')
    print(f'"{inputfile}": Answer to Part 2 = {Part2(inputfile)}')

if __name__ == "__main__":
    main()

File: AdventOfCode2022/Go/Day11/Day11.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Monkey struct {
	number                 int
	items                  []int
	operation              string
	divisible_test         int
	if_test_true_throw_to  int
	if_test_false_throw_to int
	inspections            int
}

func Parse(filename string) []Monkey {

	file, _ := os.ReadFile(filename)

	data := strings.Split(string(file), "\n")

	monkeys := []Monkey{}
	currentMonkey := Monkey{}

	for i, _ := range data {

		line := strings.TrimSpace(data[i])
		first4 := ""

		if line != "" {
			first4 = line[:4]
		} else {
			first4 = ""
		}

		switch first4 {

		//new monkey
		case "Monk":
			currentMonkeyNumber, _ := strconv.Atoi(strings.Split(line, " ")[1][:1])
			currentMonkey = Monkey{number: currentMonkeyNumber}

		// starting items
		case "Star":
			items := strings.Split(strings.Split(line, ": ")[1], ", ")
			itemsInts := []int{}
			for _, item := range items {
				itemInt, _ := strconv.Atoi(item)
				itemsInts = append(itemsInts, itemInt)
			}
			currentMonkey.items = itemsInts

		// operation
		case "Oper":
			op := strings.Split(strings.Split(line, ": ")[1], "new = old ")[1]
			currentMonkey.operation = op

		// test case
		case "Test":
			test := strings.Split(strings.Split(line, ": ")[1], "y ")[1]
			testInt, _ := strconv.Atoi(test)
			currentMonkey.divisible_test = testInt

		// true test case
		case "If t":
			testTrue := strings.Split(strings.Split(line, ": ")[1], "y ")[1]
			testTrueInt, _ := strconv.Atoi(testTrue)
			currentMonkey.if_test_true_throw_to = testTrueInt

		// false test case
		case "If f":
			testFalse := strings.Split(strings.Split(line, ": ")[1], "y ")[1]
			testFalseInt, _ := strconv.Atoi(testFalse)
			currentMonkey.if_test_false_throw_to = testFalseInt

		//end of monkey
		case "":
			monkeys = append(monkeys, currentMonkey)
			continue
		}

	}
	// get final monkey
	monkeys = append(monkeys, currentMonkey)

	return monkeys
}

func Solve(monkeys []Monkey, rounds int, decreaseWorry func(int, []Monkey) int, debug bool) int {
	
	for r := 0; r < rounds; r++ {
		if debug {
			fmt.Print("\n\nROUND ", r)
		}
		for _, monkey := range monkeys {
			for _, item := range monkey.items {

				// opersation - messy as no eval function in go
				operation := monkey.operation
				if strings.Contains(monkey.operation, "old") {
					operation = strings.ReplaceAll(operation, "old", strconv.Itoa(item))
				}

				coeficent, _ := strconv.Atoi(operation[2:])

				switch operation[:1] {
				case "+":
					item = item + coeficent
				case "-":
					item = item - coeficent
				case "*":
					item = item * coeficent
				case "/":
				}

				// decrease worry
				item = decreaseWorry(item, monkeys)

				// test
				if item%monkey.divisible_test == 0 {
					monkeys[monkey.if_test_true_throw_to].items = append(monkeys[monkey.if_test_true_throw_to].items, item)

				} else {
					monkeys[monkey.if_test_false_throw_to].items = append(monkeys[monkey.if_test_false_throw_to].items, item)
				}
				monkeys[monkey.number].inspections++
			}
			monkeys[monkey.number].items = []int{}
		}
		if debug {
			for _, monkey := range monkeys {
				fmt.Print("\nMonkey ", monkey.number, ": ")
				for _, item := range monkey.items {
					fmt.Print(item, " ")
				}
			}
		}
	}

	// calculate monkey business
	monkeysInspections := []int{}
	for _, monkey := range monkeys {
		monkeysInspections = append(monkeysInspections, -monkey.inspections)
	}
	sort.Ints(monkeysInspections[:])
	monkeyBusiness := monkeysInspections[0] * monkeysInspections[1]

	return monkeyBusiness
}

func DecreaseWorryDivide(item int, monkeys []Monkey) int {
	return int(item / 3)
}

func DecreaseWorryLCM(item int, monkeys []Monkey) int {
	LCM := 1
	for _, monkey := range monkeys {
		LCM *= monkey.divisible_test
	}
	return int(item % LCM)
}

func main() {
	
	fmt.Println("Advent-Of-Code 2022 - Day10")
	fmt.Printf("Tests : Answer to Part 1 = %v\n", Solve(Parse("tests.txt"),  20,    DecreaseWorryDivide, false))
	fmt.Printf("Inputs: Answer to Part 1 = %v\n", Solve(Parse("inputs.txt"), 20,    DecreaseWorryDivide, false))
	fmt.Printf("Tests : Answer to Part 2 = %v\n", Solve(Parse("tests.txt"),  10000, DecreaseWorryLCM,    false))
	fmt.Printf("Inputs: Answer to Part 2 = %v\n", Solve(Parse("inputs.txt"), 10000, DecreaseWorryLCM,    false))

}
